apps/charterafrica/src/lib/tools/processData.js

```diff
@@ -0,0 +1,157 @@
+  };
+
+  const languagesTechSkills = data.languages?.nodes?.map((language) => ({
+    language: language?.name,
+  }));
+  const tool = {
+    externalId,
+    avatarUrl: data?.openGraphImageUrl ?? null,
+    name: data?.name ?? null,
+    description,
+    link: data?.url ?? null,
+    location: data?.location ?? null,
+    topic,
+    languagesTechSkills,
+    lastCommit: data?.defaultBranchRef,
+    stars: data?.stargazers?.totalCount,
+    views: data?.watchers?.totalCount,
+    forks: data?.forks?.totalCount,
+    people,
+    organisation,
+    source: "github",
+  };
+  return tool;
+};
+
+const processSheet = async (update = false) => {
+  const {
+    columnMappings: {
+      toolLink,
+      toolDescription,
+      toolName,
+      toolTopic,
+      toolLocation,
+    },
+  } = await api.findGlobal(DIGITAL_DEMOCRACY_ECOSYSTEM, {});
+  const data = mockData;
+  const uniqueEntries = Object.values(
+    data.reduce((acc, obj) => {
+      acc[obj[toolLink]] = obj;
+      return acc;
+    }, {})
+  );
+
+  const toProcess = uniqueEntries.map(async (rawData, i) => {
+    const externalId = rawData[toolLink]
+      ?.replace(/^https?:\/\/github\.com\//, "")
+      .replace(/\/$/, "");
+    let [repositoryOwner, repositoryName] = externalId.split("/");
+    repositoryOwner = repositoryOwner?.trim();
+    repositoryName = repositoryName?.trim();
+    if (repositoryName && repositoryOwner) {
+      const dataFromSheet = {
+        externalId,
+        name: rawData[toolName],
+        description: rawData[toolDescription],
+        location: rawData[toolLocation],
+        topic: rawData[toolTopic],
+      };
+      let savedData;
+      const { docs } = await api.getCollection(TOOL_COLLECTION, {
+        where: {
+          externalId: { equals: externalId },
+        },
+      });
+      if (docs.length) {
+        savedData = docs?.[0];
+        if (!update) {
+          return savedData;
+        }
+      }
+      const gitData = await fetchRepository({
+        repositoryOwner,
+        repositoryName,
+      });
+      const toCreate = processRepository(gitData, dataFromSheet);
+      if (update) {
+        if (savedData) {
+          const tool = await updateOrCreateTool({
+            ...toCreate,
+            id: savedData.id,
+          });
+          return tool;
+        }
+        throw new FetchError(`Data does not exist`, rawData, 500);
+      }
+      const tool = await updateOrCreateTool(toCreate);
+      return tool;
+    }
+    throw new FetchError(
+      `Tool is invalid at row ${i}. Use format *CodeForAfrica/ui*`,
+      rawData,
+      500
+    );
+  });
+  const promises = await Promise.allSettled(toProcess);
+  const fulfilled = promises
+    .filter((p) => p.status === "fulfilled")
+    .map((p) => p.value);
+  const rejected = promises.filter((p) => p.status === "rejected");
+
+  return { fulfilled, rejected };
+};
+
+const processData = async (req, res) => {
+  try {
+    const { update } = req.query;
```

Comment:
Why do we need to specify `update` or `create`? If I understand everything all we need is a sync operation i.e.
```
Read spreadsheet/Airtable -> fetching corresponding data from GitHub/GitLab/etc., store updated content in the collections.
```
This means:
1. New items in the gSheet/Airtable but not in collections -> create them
2. Existing items in both gSheet/Airtable and collections -> update them
3. Removed items from gSheet/Airtable but exists in collections -> Ideally we should delete them but we can also have a field in the collection e.g. deleted to indicate that this item shouldn't be sent to the front end.

I don't think we need to manually specify update or create.

Reply:
Create hourly cron reads **new entries** from Airtable/Gsheet then processes by making Api calls to github.
Update cron(We do not expect this to be hourly probably daily) on the other hand will run through all the records to update fields that may have changed such as stars count, views count, edited description and tools.

The function seems to be similar so we could only need to have a parameter to specify the operation we need.

Reply:
Got you. It may be efficient to have these 2 processes separate.

But before we commit to this, did you check if GraphQL supports conditional requests similar to [REST](https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#conditional-requests)? Conditional requests don't count towards rate limit and hence a single cron job can run and handle everything for us.